Minimal initial func.yaml

## Summary

All fields on the `Function` struct should be set to `omitempty`

## Exposition

When a user creates a new Function, their `func.yaml` should include only the values which are either required (language), have no default (created date/time stamp) or deviate from static defaults (`func create -l go -t cl.

With the current version of the functions core library, this would _ideally_ look as such:
```shell
>  mkdir f && cd f && func create -l go
Created go function in [path]

> cat func.yaml
specVersion: 0.25.0
runtime: go
created: 2022-09-26T23:06:48.442316+09:00
```

## Alternative Considered: The fully-populated Config File approach
 
A traditional "config file" style setup would be where the func.yaml is fully populated with all possible values (commented and showcasing static defaults), whereby the user can choose to uncomment a given setting.  However, this usage pattern is directly at odds with the usage pattern we are encouraging: CLI or direct API.  For example, we using our friend `git` as an example, only when configuration is changed using `git config ...` does the `.git/config` file change.  This is the opposite approach to, for example `sshd`, which upon installation provides a comprehensive config file available at `/etc/ssh/sshd_config`.  The reason for the difference is that the former has a CLI which provides for configuration, while the latter does not.  We have a CLI.

This is of course up for debate.  It might be nice to account for a user who prefers  a `func.yaml`-first approach.  That may very well happen in the future, but this minimal approach should probably still be the first step.  We currently have the worst of both worlds, were func.yaml is populated with an intimidating litany of uncommented zero values.

## Caviat: It won't be that simple right now

Even if all fields are set to `omitempty`, the func.yaml will still include several more values.  This is a result of having not been strict in our usage of in-code static defaults early on, and requiring a few values we don't really need, so there are still a few fields which are written to every func.yaml despite not being actually deviations from the default.  The closest we can get _at present_ would be:

```
specVersion: 0.25.0
name: f
runtime: go
healthEndpoints:
  liveness: /health/liveness
  readiness: /health/readiness
buildpacks:
- paketo-buildpacks/go-dist
- ghcr.io/boson-project/go-function-buildpack:tip
created: 2022-09-26T23:06:48.442316+09:00
invocation:
  format: http
```

## Related

See forthcoming issues which request that we rely on the static defaults for buildpacks and health endpoints, only require a name upon deploy, and flatten the `invocation` struct (also relying on `http` as the static default)

Hi @lkingland I added the flag and created a PR. However, during the CI, there were two errors when deploying to cluster. One caused by the `Image` and the other `Registry`. I removed the `omitempty` from those two and it all worked. Currently the PR passes CI. The errors are below:

Image:
```
=== RUN   TestBasicUpload
    cmd.go:83: /home/runner/work/func/func/func create -l node /tmp/TestBasicUpload3227201013/001/test-func-basic-upload29khr
    cmd.go:83: /home/runner/work/func/func/func deploy -p /tmp/TestBasicUpload3227201013/001/test-func-basic-upload29khr -r ttl.sh/knfunc2pqk62fp --remote --verbose
    cmd.go:127: Creating Pipeline resources
        Running Pipeline Task: Building function image on the cluster
        Running Pipeline Task: Deploying function to the cluster
        Error: failed to run pipeline: function pipeline run has failed with message: 
        
        ⬆️  Deploying function to the cluster
        deploy error: knative deployer failed to deploy the Knative Service: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: missing field(s): spec.template.spec.containers[0].image
        Error: knative deployer failed to deploy the Knative Service: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: missing field(s): spec.template.spec.containers[0].image
        
        
    cmd.go:128: exit status 1
    cmd.go:83: /home/runner/work/func/func/func invoke -p /tmp/TestBasicUpload3227201013/001/test-func-basic-upload29khr
    cmd.go:127: Error: services.serving.knative.dev "test-func-basic-upload29khr" not found
        
    cmd.go:128: exit status 1
    scenario_basic_test.go:41: assertion failed: expression is false: strings.Contains(result.Out, "first revision"): Func body does not contain 'first revision'
    cmd.go:83: /home/runner/work/func/func/func delete -p /tmp/TestBasicUpload3227201013/001/test-func-basic-upload29khr
    cmd.go:127: 
        
   Removing Knative Service: test-func-basic-upload29khr
        
   Removing Knative Service 'test-func-basic-upload29khr' and all dependent resources
        Error: knative remover failed to delete the service: services.serving.knative.dev "test-func-basic-upload29khr" not found
        
    cmd.go:128: exit status 1
--- FAIL: TestBasicUpload (76.11s)
```
Registry, after removing `omitempty` from Image:
```
=== RUN   TestBasicGit
    gitserver.go:83: Initialized HTTP Func Git Server: Server URL = http://func-git.default.127.0.0.1.sslip.io/ Pod Name = func-git-00001-deployment-69ff558-r54wv
    cmd.go:83: kubectl exec func-git-00001-deployment-69ff558-r54wv -c user-container -- git-repo create test-func-basic-gitctx2s
    cmd.go:122: Initialized empty Git repository in /var/www/git/test-func-basic-gitctx2s.git/
        created
    cmd.go:83: /home/runner/work/func/func/func create -l node /tmp/TestBasicGit4213784320/001/test-func-basic-gitctx2s
    cmd.go:83: sh -c git init
    cmd.go:83: sh -c git branch -M main
    cmd.go:83: sh -c git add .
    cmd.go:83: sh -c git config user.name "John Smith"
    cmd.go:83: sh -c git config user.email "john.smith@example.com"
    cmd.go:83: sh -c git commit -m "initial commit"
    cmd.go:83: sh -c git remote add origin http://func-git.default.127.0.0.1.sslip.io/test-func-basic-gitctx2s.git
    cmd.go:83: sh -c git push -u origin main
    cmd.go:83: /home/runner/work/func/func/func deploy -p /tmp/TestBasicGit4213784320/001/test-func-basic-gitctx2s -r ttl.sh/knfuncmhujvp0s --remote --verbose --git-url http://func-git.default.svc.cluster.local/test-func-basic-gitctx2s.git
    cmd.go:127: Creating Pipeline resources
        Running Pipeline Task: Fetching git repository with the function source code
        Running Pipeline Task: Building function image on the cluster
        Running Pipeline Task: Deploying function to the cluster
        Error: failed to run pipeline: function pipeline run has failed with message: 
        
        Warning: function has current image 'ttl.sh/knfuncmhujvp0s/test-func-basic-gitctx2s:latest' which has a different registry than the currently configured registry 'image-registry.openshift-image-registry.svc:5000/default'. The new image tag will be 'image-registry.openshift-image-registry.svc:5000/default/test-func-basic-gitctx2s:latest'.  To use an explicit image, use --image.
        ⬆️  Deploying function to the cluster
        Waiting for Knative Service to become ready
        deploy error: knative deployer failed to wait for the Knative Service to become ready: RevisionFailed: Revision "test-func-basic-gitctx2s-00001" failed with message: Unable to fetch image "image-registry.openshift-image-registry.svc:5000/default/test-func-basic-gitctx2s:latest": failed to resolve image to digest: Get "https://image-registry.openshift-image-registry.svc:5000/v2/": dial tcp: lookup image-registry.openshift-image-registry.svc on 10.96.0.10:53: no such host.
        Error: knative deployer failed to wait for the Knative Service to become ready: RevisionFailed: Revision "test-func-basic-gitctx2s-00001" failed with message: Unable to fetch image "image-registry.openshift-image-registry.svc:5000/default/test-func-basic-gitctx2s:latest": failed to resolve image to digest: Get "https://image-registry.openshift-image-registry.svc:5000/v2/": dial tcp: lookup image-registry.openshift-image-registry.svc on 10.96.0.10:53: no such host.
        
        
    cmd.go:128: exit status 1
    cmd.go:83: /home/runner/work/func/func/func invoke -p /tmp/TestBasicGit4213784320/001/test-func-basic-gitctx2s
    cmd.go:127: Error: failure invoking 'http://test-func-basic-gitctx2s.default.127.0.0.1.sslip.io/' (HTTP 404)
        
    cmd.go:128: exit status 1
    scenario_basic_test.go:99: assertion failed: expression is false: strings.Contains(result.Out, "first revision"): Func body does not contain 'first revision'
    cmd.go:83: /home/runner/work/func/func/func delete -p /tmp/TestBasicGit4213784320/001/test-func-basic-gitctx2s
    cmd.go:83: kubectl exec func-git-00001-deployment-69ff558-r54wv -c user-container -- git-repo delete test-func-basic-gitctx2s
    cmd.go:122: deleted
--- FAIL: TestBasicGit (55.92s)
```

--- a/pkg/functions/function.go
+++ b/pkg/functions/function.go
@@ -29,10 +29,10 @@ type Function struct {
 	Root string `yaml:"-"`
 
 	// Name of the function.
-	Name string `yaml:"name" jsonschema:"pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"`
+	Name string `yaml:"name,omitempty" jsonschema:"pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"`
 
 	// Runtime is the language plus context.  nodejs|go|quarkus|rust etc.
-	Runtime string `yaml:"runtime"`
+	Runtime string `yaml:"runtime,omitempty"`
 
 	// Template for the function.
 	Template string `yaml:"-"`
@@ -53,7 +53,7 @@ type Function struct {
 	Image string `yaml:"image"`
 
 	// SHA256 hash of the latest image that has been built
-	ImageDigest string `yaml:"imageDigest"`
+	ImageDigest string `yaml:"imageDigest,omitempty"`
 
 	// Created time is the moment that creation was successfully completed
 	// according to the client which is in charge of what constitutes being
@@ -65,13 +65,13 @@ type Function struct {
 	Invoke string `yaml:"invoke,omitempty"`
 
 	//BuildSpec define the build properties for a function
-	Build BuildSpec `yaml:"build"`
+	Build BuildSpec `yaml:"build,omitempty"`
 
 	//RunSpec define the runtime properties for a function
-	Run RunSpec `yaml:"run"`
+	Run RunSpec `yaml:"run,omitempty"`
 
 	//DeploySpec define the deployment properties for a function
-	Deploy DeploySpec `yaml:"deploy"`
+	Deploy DeploySpec `yaml:"deploy,omitempty"`
 }
 
 // BuildSpec
@@ -88,46 +88,46 @@ type BuildSpec struct {
 	BuilderImages map[string]string `yaml:"builderImages,omitempty"`
 
 	// Optional list of buildpacks to use when building the function
-	Buildpacks []string `yaml:"buildpacks"`
+	Buildpacks []string `yaml:"buildpacks,omitempty"`
 
 	// Builder is the name of the subsystem that will complete the underlying
 	// build (pack, s2i, etc)
-	Builder string `yaml:"builder" jsonschema:"enum=pack,enum=s2i"`
+	Builder string `yaml:"builder,omitempty" jsonschema:"enum=pack,enum=s2i"`
 
 	// Build Env variables to be set
-	BuildEnvs []Env `yaml:"buildEnvs"`
+	BuildEnvs []Env `yaml:"buildEnvs,omitempty"`
 }
 
 // RunSpec
 type RunSpec struct {
 	// List of volumes to be mounted to the function
-	Volumes []Volume `yaml:"volumes"`
+	Volumes []Volume `yaml:"volumes,omitempty"`
 
 	// Env variables to be set
-	Envs []Env `yaml:"envs"`
+	Envs []Env `yaml:"envs,omitempty"`
 }
 
 // DeploySpec
 type DeploySpec struct {
 	// Namespace into which the function is deployed on supported platforms.
-	Namespace string `yaml:"namespace"`
+	Namespace string `yaml:"namespace,omitempty"`
 
 	// Remote indicates the deployment (and possibly build) process are to
 	// be triggered in a remote environment rather than run locally.
-	Remote bool `yaml:"remote"`
+	Remote bool `yaml:"remote,omitempty"`
 
 	// Map containing user-supplied annotations
 	// Example: { "division": "finance" }
-	Annotations map[string]string `yaml:"annotations"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
 
 	// Options to be set on deployed function (scaling, etc.)
-	Options Options `yaml:"options"`
+	Options Options `yaml:"options,omitempty"`
 
 	// Map of user-supplied labels
-	Labels []Label `yaml:"labels"`
+	Labels []Label `yaml:"labels,omitempty"`
 
 	// Health endpoints specified by the language pack
-	HealthEndpoints HealthEndpoints `yaml:"healthEndpoints"`
+	HealthEndpoints HealthEndpoints `yaml:"healthEndpoints,omitempty"`
 }
 
 // HealthEndpoints specify the liveness and readiness endpoints for a Runtime

--- a/pkg/functions/function_envs.go
+++ b/pkg/functions/function_envs.go
@@ -9,7 +9,7 @@ import (
 
 type Env struct {
 	Name  *string `yaml:"name,omitempty" jsonschema:"pattern=^[-._a-zA-Z][-._a-zA-Z0-9]*$"`
-	Value *string `yaml:"value"`
+	Value *string `yaml:"value,omitempty"`
 }
 
 func (e Env) String() string {

--- a/pkg/functions/function_volumes.go
+++ b/pkg/functions/function_volumes.go
@@ -5,7 +5,7 @@ import "fmt"
 type Volume struct {
 	Secret    *string `yaml:"secret,omitempty" jsonschema:"oneof_required=secret"`
 	ConfigMap *string `yaml:"configMap,omitempty" jsonschema:"oneof_required=configmap"`
-	Path      *string `yaml:"path"`
+	Path      *string `yaml:"path,omitempty"`
 }
 
 func (v Volume) String() string {

--- a/schema/func_yaml-schema.json
+++ b/schema/func_yaml-schema.json
@@ -3,11 +3,6 @@
 	"$ref": "#/definitions/Function",
 	"definitions": {
 		"BuildSpec": {
-			"required": [
-				"buildpacks",
-				"builder",
-				"buildEnvs"
-			],
 			"properties": {
 				"git": {
 					"$schema": "http://json-schema.org/draft-04/schema#",
@@ -52,14 +47,6 @@
 			"description": "BuildSpec"
 		},
 		"DeploySpec": {
-			"required": [
-				"namespace",
-				"remote",
-				"annotations",
-				"options",
-				"labels",
-				"healthEndpoints"
-			],
 			"properties": {
 				"namespace": {
 					"type": "string",
@@ -102,9 +89,6 @@
 			"description": "DeploySpec"
 		},
 		"Env": {
-			"required": [
-				"value"
-			],
 			"properties": {
 				"name": {
 					"pattern": "^[-._a-zA-Z][-._a-zA-Z0-9]*$",
@@ -120,15 +104,9 @@
 		"Function": {
 			"required": [
 				"specVersion",
-				"name",
-				"runtime",
 				"registry",
 				"image",
-				"imageDigest",
-				"created",
-				"build",
-				"run",
-				"deploy"
+				"created"
 			],
 			"properties": {
 				"specVersion": {
@@ -292,10 +270,6 @@
 			"type": "object"
 		},
 		"RunSpec": {
-			"required": [
-				"volumes",
-				"envs"
-			],
 			"properties": {
 				"volumes": {
 					"items": {
@@ -348,9 +322,6 @@
 			"type": "object"
 		},
 		"Volume": {
-			"required": [
-				"path"
-			],
 			"properties": {
 				"secret": {
 					"type": "string"

